design/k8s-job-hook.md

```diff
@@ -0,0 +1,233 @@
+
+```yaml
+hooks:
+  k8s-job:
+    allowed_images:
+      - rclone/rclone:1.57
+      - alpine
+```
+
+Each item list will match the `registry/name:tag` used in the hook.
+If the tag is omitted, any tag will be allowed.
+
+
+### Execution
+
+lakeFS requests job creation from the cluster.  In case the request fails it will log the job information and fail the hook execution.
+By default the hook waits for the job execution to complete, and based on the job status consider the execution to be successful.
+It will capture the job output into the step output.
+In case we would not like to block the operation (merge/commit) until the job execution is completed, for example on post-* events. We can change the default behavior by setting wait_for_complete to false:
+
+```yaml
+properties:
+  wait_for_complete: false
+```
+
+Without waiting for the job to complete, lakeFS will consider job creation as successful execution. Will log the job information without waiting for the job to complete or capturing the output.
+
+### Authorizations
+
+Base on the above, lakeFS will require the following permissions:
+
+- `job` get, create and watch
+- `pod` get
+- `pod/log` get, list, watch
+
+The following describes possible ClusteRole that enables the above. We can limit the scope to a single namespace using Role.
+Note that we need to add the rules to the current set used by lakefs, this document describes the requirements for this feature.
+
+```
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  namespace: default
+  name: lakefs
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: lakefs
+rules:
+  - apiGroups: [""]
+    resources: ["job"]
+    verbs: ["get", "create", "watch"]
+  - apiGroups: [""]
+    resources: ["pod"]
+    verbs: ["get"]
+  - apiGroups: [""]
+    resources: ["job/logs"]
+    verbs: ["get", "list", "watch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: lakefs
+subjects:
+  - kind: ServiceAccount
+    namespace: default
+    name: lakefs
+    apiGroup: ""
+roleRef:
+  kind: ClusterRole
+  name: lakefs
+  apiGroup: rbac.authorization.k8s.io
+```
+
+
+### Considerations
+
+*Job lifetime*
+
+Once a job is created and executed in the cluster, the lakeFS server will not take ownership of the object. A mechanism should be in place to clean up all jobs lakeFS applied and completed (successfully or not).
+[Automatic Clean-up for Finished Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/) capability is currently found on Kubernetes 1.23 (which we don’t have yet on AWS for example) which can help with that.
+
+*Job security*
+
+A user-supplied image can be run within the scope specified in the job spec, but lakeFS has no control over the job and the code itself, except for the image name that is limited by the lakeFS configuration.
```

Comment:
As above -- I would not regard whitelisting images a usable security mechanism in most environments.  Instead, we should hook into the existing K8s permissions structure, and just request a configurable role to run the job.

Reply:
Rewrote this section, but I didn't removed the whitelist part yet because I didn't find any solution inside the cluster to limit the execution.